### ISSUE TITLE
Deserialize custom errors

### ISSUE THREAD

**Issue by geigerj** [googleapis/gapic-generator#1357]
### Context

Original report was at https://github.com/GoogleCloudPlatform/google-cloud-ruby/issues/1438

### What

Google APIs can return custom error types. These custom types are represented as proto messages in the API. These types are not currently available on GitHub.

### How

gRPC returns a serialized instance of the custom error type as ~~an `Any` proto~~ a proto. These types need to be made available on GitHub. Each language's GAX should then deserialize that ~~`Any`~~ proto into the correct type before wrapping it in a `GaxError`.

EDIT: the error returned in trailing metadata seems to be a serialized instance of the error proto itself, not an `Any` instance.

- [x] open source error types

#### Update GAX

- [ ] Java
- [ ] Python
- [ ] C#
- [x] Ruby
- [x] Go
- [ ] PHP
- [ ] NodeJS

**Comment by geigerj** [comment]
cc: @qingling128

**Comment by jba** [comment]
gax for Go does not try to interpret the errors returned by gRPC. Users can always use the [grpc/status](https://godoc.org/google.golang.org/grpc/status) package to access the parts of the error.

Should we be doing more? Presumably we'd need another type that wrapped `grpc.status.Status`.

**Comment by geigerj** [comment]
@jba Right -- the problem is that users can access the parts of the error, but those parts might be serialized protos representing custom errors. The user cannot easily deserialize the proto because the custom error proto definitions themselves are not open sourced. Even if users could, I think it's a bad user experience to ask users to deserialize the error details if we're able to do that for them.

We should have something that looks like `google.rpc.Status`, but where [`Status.details`](https://github.com/googleapis/googleapis/blob/1e724373436dc016f1708dd18c906524532716ea/google/rpc/status.proto#L91), instead of being a serialized `Any`, is the wrapped type.

The reason this conversion falls on GAPIC rather than gRPC is that gRPC has no knowledge of which proto types are being represented in that `Any` by the API, whereas we do know about these types. (Or at least we will, once they are open sourced.)

**Comment by pongad** [comment]
To make sure I'm understanding this right: The custom error is API specific right? Say, method `Foo` might return error with details typed `AErrorDetail` and method `Bar` might have `BErrorDetail`. The issue is that we should deserialize the Detail into either `AErrorDetail` or `BErrorDetail` (I assume we'll know which error it is by some config)? First we need to open source both detail types?

**Comment by geigerj** [comment]
> The custom error is API specific right? Say, method Foo might return error with details typed AErrorDetail and method Bar might have BErrorDetail. The issue is that we should deserialize the Detail into either AErrorDetail or BErrorDetail

Correct, although the errors are API- (not necessarily method-) specific.

> (I assume we'll know which error it is by some config)

We'll know which error it is from the [`type_url` field of the original `detail`](https://github.com/google/protobuf/blob/master/src/google/protobuf/any.proto#L135). (Annoyingly, in the original issue report that sparked this, the user could therefore see the name of the proto message, but the message definition itself was nowhere to be found in open source...)

> First we need to open source both detail types?

Right.

**Comment by jba** [comment]
@pongad, we don't have to worry about the types. We just use `ptypes.UnmarshalAny`. The problem is that that won't work unless the right proto is linked in to the binary.

So we need to know what to link in, if the messages aren't part of the service protos. (They should be, but they don't have to be, and you'll never know until you try to access error details.)

We also might want to make it easier to retrieve the details. I would not want to add a third status type on top of the proto and the gRPC status type. But we could either provide our own `StatusDetails(*status.Status) []proto.Message` method, or  lobby gRPC to add a `Status.Details() []proto.Message` method to the `status` package.

**Comment by pongad** [comment]
@jba You're right. I didn't realize the `details` field is repeated. I was thinking about wrapping the error earlier, but that would mess with `status.FromError`.

I think it's probably best to lobby gRPC. I'll see what I can do :smile:

**Comment by qingling128** [comment]
Hi @pongad, any update on this one?

**Comment by pongad** [comment]
@qingling128 Go is being tracked by https://github.com/grpc/grpc-go/pull/1358 . @ghasemloo beat me to the issue :)

**Comment by qingling128** [comment]
Cool cool. Thanks for checking! Is the change specific to Go blocking the open sourcing? Or are they two separate things? Just wondering which one I should track.

**Comment by pongad** [comment]
No problem! I wasn't being very clear; let me back up. IIUC, there are two pieces to this problem.

1. We need to open source the error protos. Otherwise, we don't know how to deserialize the errors

2. Each language needs to provide a way to get to the deserialized error.

Both are required, but they can be implemented separately. The issue I linked will fix (2) for Go. Am I making any sense?

**Comment by qingling128** [comment]
Yup, that makes total sense! Thank you so much for the clarification.

BTW, do we have a rough idea how long the open sourcing process like this takes in general? Is it like days, weeks or over a month? (Trying to take this into consideration with our Q3 planning. This has been blocking one of our P0 requests for a month. If it takes a while, we might need to seek for some workaround.)

Thanks,
Ling

**Comment by pongad** [comment]
I don't really know. I believe @geigerj is the one open sourcing this.

**Comment by qingling128** [comment]
Cool cool. Thanks anyway!

@geigerj - Any chance you know the answer to the question above?

**Comment by geigerj** [comment]
@qingling128 The open sourcing process generally takes on the order of days, almost entirely toward setting up some tooling and going through code reviews inside Google. Once that's done, the code is synchronized automatically to GitHub.

I'll update in the next day or so with an ETA.

**Comment by qingling128** [comment]
Glad to hear that. Thanks @geigerj!

**Comment by geigerj** [comment]
@eoogbe will be working on open sourcing the custom error types for Logging in the next week or two.

**Comment by eoogbe** [comment]
[They are already in googleapis](https://github.com/googleapis/googleapis/blob/master/google/logging/v2/logging.proto#L152)

**Comment by qingling128** [comment]
I vaguely remember we reviewed these protos in the meeting and figured there were still something missing. @geigerj might have a better idea than I do.

**Comment by geigerj** [comment]
I think the message that Eva linked to is the one that's needed.

(I also can't recall exactly how we determined that there was more open sourcing needed during our last meeting -- that might have been a mistake. It looks like the message was open sourced after the initial bug report but before our meeting, which might have led to some confusion.)

**Comment by qingling128** [comment]
Thanks for the clarification! So we are unblocked for that part.

Is there additional work needed in `GAX` for Ruby support of custom errors?

**Comment by geigerj** [comment]
Right, we still need to update `GAX` to deserialize the error.

**Comment by qingling128** [comment]
Sounds good! Is that tracked in this ticket as well? Any rough ETA?

**Comment by geigerj** [comment]
@qingling128 Yes, that's tracked in this ticket, too. I think we should be able to get a change out in GAX next week.

**Comment by qingling128** [comment]
Sounds great. Thanks for the heads up!

**Comment by pongad** [comment]
Since Go is already finished (I take no credit, since I didn't do anything), I'll un-assign myself. Please let me know if anyone needs anything.

**Comment by qingling128** [comment]
@geigerj - Hi, does it make sense to have this assigned back to you? Feel free to re-assign if necessary. Thanks!

**Comment by jskeet** [comment]
It's unclear to me whether C# needs anything for this or not. The exception is thrown by Grpc.Core, and we don't intercept it.

If the error protos are public, then users who wish to *can* deserialize appropriately, although currently they need to specify which type to *try* to deserialize as. Maybe the simplest next step is to take a concrete example and show what the client code looks like.

**Comment by dklimkin** [comment]
@jskeet Ideally, I'd like to specify message name in the apiname.yml or apiname_gapic.yml and have toolkit intercept the grpc core error, deserialize my message and wrap in a ApiNameException with details filed of this message.

**Comment by jskeet** [comment]
@dklimkin: Changing the exception that's thrown for all APIs is a *big* change at this point (especially given that some APIs have shipped). I see the point, but I'm not sure that giving every API its own exception type is a good idea either. (Deriving the new exception from the currently-thrown exception type would be reasonable though... will consider that more.)

**Comment by dklimkin** [comment]
Extending the existing one should solve issues with the already shipped code.

As per per-API exceptions, consider a user code that uses multiple APIs simultaneously. They would need to distinguish which of the calls failed -- and create API-specific exceptions on the client side anyway.

**Comment by jskeet** [comment]
I've confirmed that the gRPC `RpcException` in C# is unsealed, so that part should be okay. Considering it at the moment. (But we have a *ton* of other work going on as well...)

**Comment by qingling128** [comment]
Hi @geigerj, any chance you've got some time to look into the GAX lib change in Ruby yet this week?

**Comment by geigerj** [comment]
Hi @qingling128 -- sorry, I haven't had a chance to make the GAX change yet. It is on my radar and I will update once it's done.

**Comment by qingling128** [comment]
Cool cool. Let me know when it's out. I could help test the change. Thanks!

**Comment by pongad** [comment]
Looks like Go might not actually be done after all. We need to make sure to link in the error types, otherwise we cannot serialize.
@geigerj is there a way to tell what types the errors could be? I thought it was in YAMLs somewhere but I can't seem to find it....

**Comment by geigerj** [comment]
@pongad I don't think they're in the YAMLs. The types should be defined in the APIs' source protos ([example](https://github.com/googleapis/googleapis/blob/f781f331a0d0269725d33c715e68fae9129f0214/google/logging/v2/logging.proto#L152)), so they should be automatically included in the protoc-generated library.

As far as determining which error type is being returned so that you can deserialize, I'm finding harder. It looks like I was incorrect in my original report when I claimed that the error message is returned as an instance of `Any`. For Ruby, it's actually returned as a serialized instance of the proto itself, which makes it difficult to know how to deserialize it. It's possible the protoc-generated gRPC client libraries in the other languages do some magic to make this easier, but I'm not sure. I started an internal thread to discuss for Ruby, and will add you to that in case any of the context is relevant to the Go work.

**Comment by qingling128** [comment]
Hi @geigerj, we are expecting b/65028199 to be fixed by the end of this week. If that's possible, how long do you think would the remaining work take to expose the deserialized details in the GAX error (at least for Ruby if it needs to be done per language)?

**Comment by geigerj** [comment]
Hi @qingling128, Assuming the fix for that internal bug results in an accessor for an instance of `status.proto` (as discussed in an offline thread), I should be able to update GAX within a few days of the fix. This work will need to be done per-langauge.

**Comment by qingling128** [comment]
Thanks @geigerj! Will let you know as soon as that internal bug is fixed. We are observing some significant amount of dropped logs (b/65328414). It would be highly appreciated if we could enable partial success in the Logging Agent asap. Right now we have no way to tell apart "good" and "bad" log entries, so we end up retrying over and over, which ultimately drops "good" entries as well and crashes the agent.

**Comment by geigerj** [comment]
@qingling128 Acknowledged. I will keep you updated after the internal bug unblocks this work.

**Comment by qingling128** [comment]
Great, thanks!

**Comment by qingling128** [comment]
Hi @geigerj, great news! The internal bug has just been fixed. Could you help update Ruby GAX accordingly?

**Comment by geigerj** [comment]
Done in Ruby GAX at https://github.com/googleapis/gax-ruby/pull/89, https://github.com/googleapis/gax-ruby/pull/93

**Comment by pongad** [comment]
Done for Go by gRPC layer.

**Comment by vchudnov-g** [comment]
@pongad @lukesneeringer @garrettjonesgoogle Is this still an issue for the remaining languages?

**Comment by vchudnov-g** [comment]
Friendly ping! Need an update to meet out SLOs.  Is this still an issue for the remaining languages?

**Comment by jskeet** [comment]
Still a potential nice-to-have, but I don't think we should even think about it before the microgenerators are finished.

**Comment by vchudnov-g** [comment]
SLO update? Sounds like we should keep this at P3 for now

**Comment by chrisdunelm** [comment]
gRPC interceptors now provide this functionality in all languages as far as I know.
If no-one disagrees, this issue should be closed.
